### DISCUSSION TITLE
rdk_doc/Advanced_development/hardware_development/rdk_x5/can

### DISCUSSION THREAD

**Discussion by giscus** [D-Robotics/rdk_doc#54] [General]
# rdk_doc/Advanced_development/hardware_development/rdk_x5/can

协议简介

https://d-robotics.github.io/rdk_doc/Advanced_development/hardware_development/rdk_x5/can

**Comment by ss123-dev** [comment]
使用can工具调试时候发现无法通信

**Reply by xuguosheng110** [reply]
你好，请更新到最新版本的3.1.0系统，已经对can模块进行修复

**Comment by sunyngzhng** [comment]
请问下这个文件在哪里看到？kernel\drivers\net\can\m_can\tcan4x5x-core.c 。在RDKX5 开发板运行Linux系统（Ubuntu 20.4）中中没有发现这个文件。在x5-kernel-main.zip内核源码中也没发现该文件。

**Reply by WuChao-2024** [reply]
这部分属于驱动开发章节，参考RDK X5 的驱动开发指南即可

**Comment by akipop9** [comment]
使用can过程中，通信7次左右报错，缓冲区溢出，增大缓冲区也没有效果，与开发板通信的电机can用win10调试过，确定是没问题的，下面是我原始的python代码。
import can
import time

def send_one():
    bus = can.interface.Bus(interface='socketcan',channel='can0',bitrate=1000000)
    print("init success")

    for i in range(100):

        msg = can.Message(arbitration_id=0x141,data=[0xA1,0x00,0x00,0x00,0x64,0x00,0x00,0x00],is_extended_id=False)
        
        try:
            bus.send(msg)
        except can.CanError as e:
            print(f"msg not send.{e}")
        res = bus.recv(timeout=1.0)
        time.sleep(0.1)
        msg = can.Message(arbitration_id=0x141,data=[0xA1,0x00,0x00,0x00,0x00,0x00,0x00,0x00],is_extended_id=False)
        
        try:
            bus.send(msg)
        except can.CanError:
            print("msg not send:",i)
        res = bus.recv(timeout=1.0)
        time.sleep(0.1)

    print(res)
    bus.shutdown()
send_one()
命令行显示内容：
sunrise@ubuntu:~/testPycan$ python3 ./test.py 
init success
msg not send.Failed to transmit: No buffer space available [Error Code 105]
msg not send: 7
msg not send.Failed to transmit: No buffer space available [Error Code 105]
我查找了很多资料也没有找到解决办法。

**Reply by WuChao-2024** [reply]
你好，可以烧录社区目前RDK X5最新的系统，烧录最新的miniboot来使用CAN，相关问题均已经修复。目前建议更新的方式是在可以进入Ubuntu系统中，更新miiniboot后，烧录最新系统，然后在最新系统中再更新一次miniboot，避免系统和miniboot版本差异过大导致无法启动。
https://forum.d-robotics.cc/t/topic/30925

最后，建议使用root用户来使用can相关功能

**Reply by sunyngzhng** [reply]
你好，请问下，已更新至rdk-x5-ubuntu22-preinstalled-desktop-3.5.0-arm64.img.xz。还需要再更新一次miniboot吗？minboot 和系统的关系是什么？

**Comment by lanhongbiao** [comment]
回环测试candump那边没消息是为啥

**Reply by WuChao-2024** [reply]
你好，可以烧录社区目前RDK X5最新的系统，烧录最新的miniboot来使用CAN，相关问题均已经修复。目前建议更新的方式是在可以进入Ubuntu系统中，更新miiniboot后，烧录最新系统，然后在最新系统中再更新一次miniboot，避免系统和miniboot版本差异过大导致无法启动。
https://forum.d-robotics.cc/t/topic/30925

**Comment by WangShuoran** [comment]
请问上面提到可以用CAN2.0 B进行通信是吗？需要哪些协议的修改？我这边希望使用2.0B标准帧进行数据传输，谢谢指导

**Comment by mimimamakaka** [comment]
请问怎样使用canfd接口，与can协议的电机通讯，can2.0a 上面介绍是支持的，但是没有告知具体实现方式
